### PR TITLE
add sickle module

### PR DESCRIPTION
**Pull request by noirot**
- [X] There is example tool output for tools in the https://github.com/ewels/MultiQC_TestData repository
 - [X] Code is tested and works locally (including with `--lint` flag)
 - [X] `CHANGELOG.md` is updated
 - [X] `README.md` is updated
 - [X] `docs/README.md` is updated with link to below
 - [X] `docs/modulename.md` is created
 - [X] Everything that can be represented with a plot instead of a table is a plot
 - [X] Report sections have a description and help text (with `self.add_section`)
 - [X] There aren't any huge tables with > 6 columns (explain reasoning if so)
 - [X] Each table column has a different colour scale to its neighbour, which relates to the data (eg. if high numbers are bad, they're red)
 - [X] Module does not do any significant computational work

### CODE REVIEW

**Comment by ewels** [comment]
Hi @noirot! Is there any reason that you closed this?

**Comment by noirot** [comment]
No real reason , I thought it was merged ... I'm not used with PR.

**Comment by ewels** [comment]
Ah ok - no not yet sorry. I merged the PR with the test data but I haven’t got to the code yet. Should do soon..

Phil

File: multiqc/modules/sickle/sickle.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+
+        self.sickle_data = dict()
+        #parse list of log files
+        for f in self.find_log_files('sickle'):
+            self.sickle_data[f['s_name']] = self.parse_logs(f['f'])
+            self.add_data_source(f)
+        #no file found
+        if len(self.sickle_data) == 0:
+            raise UserWarning
+        self.sickle_data = self.ignore_samples(self.sickle_data)
```
Comment (ewels) [inline multiqc/modules/sickle/sickle.py:28]: This needs to go before the above check, as if samples are ignored then we might end up with 0..

File: multiqc/modules/sickle/sickle.py [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+
+    def parse_logs(self, f):
+        """ Parse the Sickle standard output """
+        regexes = {
+                'reads_paired_kept': "FastQ paired records kept: ([\d,]+) .*",
+                'reads_single_kept': "FastQ single records kept: ([\d,]+).*",
+                'reads_paired_discarded': "FastQ paired records discarded: ([\d,]+) .*",
+                'reads_single_discarded': "FastQ single records discarded: ([\d,]+) .*",
+                'reads_kept': "FastQ records kept: ([\d,]+)",
+                'reads_discarded': "FastQ records discarded: ([\d,]+)"
```
Comment (ewels) [inline multiqc/modules/sickle/sickle.py:41]: As these search patterns are specific to single-end data, I think that we can call them `reads_single_kept` / `reads_single_discarded` (but needs a minor code refactor as we can't have duplicate keys in a dict).

File: multiqc/modules/sickle/sickle.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            data["reads_discarded"] = data["reads_paired_discarded"] + data["reads_single_discarded"]
+
+        return data
+
+    def sickle_general_stats_table(self):
+        """ Take the parsed stats from the sickle report and add
+        number of kept reads into the generam stats table """
+
+        headers = {}
+        headers['reads_kept'] = {
```
Comment (ewels) [inline multiqc/modules/sickle/sickle.py:65]: I guess that the key thing most people will want to know is what percentage of their library was removed by Sickle. I've calculated a percentage and added this to the table.

File: multiqc/modules/sickle/sickle.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        self.general_stats_addcols(self.sickle_data, headers)
+
+    def read_count_plot (self):
+        """ Stacked bar plot showing counts of reads """
+        pconfig = {
+            'id': 'sickle_sequence_counts_plot',
+            'title': 'Sickle: Sequence Counts',
+            'ylab': 'Number of reads',
+            'cpswitch_counts_label': 'Number of reads',
+            'hide_zero_cats': False
```
Comment (ewels) [inline multiqc/modules/sickle/sickle.py:79]: I've removed this as it allows us to simplify the code below a lot (we don't need to care whether samples are paired-end or not). I think that the changes of having _zero_ trimmed reads are pretty unlikely so I think it's safe enough.

File: multiqc/modules/sickle/sickle.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+        for l in f.splitlines():
+            s = l.split(':')
+            # Search regexes for overview stats# Search regexes for overview stats
+            for k, r in regexes.items():
+                match = re.search(r, l)
+                if match:
+                    data[k] = int(match.group(1).replace(',', ''))
+        #if paired data : compute total for general stats table
+        if "reads_kept" not in data.keys() :
+            data["reads_kept"] = data["reads_paired_kept"] + data["reads_single_kept"]
```
Comment (ewels) [inline multiqc/modules/sickle/sickle.py:53]: Do we need to duplicate the paired count here? I'm not sure whether we can really add pairs and singletons... 🤔

**Comment by noirot** [comment]
Thanks a lot for reviewed !

**Comment by ewels** [comment]
Hi @noirot - please reopen the pull request again :wink: You don’t need to close pull requests - when I’m happy I will merge it (as I am the maintainer for this repository). You only close a PR if you decide that actually you didn’t want to contribute code after all (unlike issues, which are closed when resolved).

Phil

**Comment by ewels** [comment]
The reason that I haven’t merged yet is because of my question about counting read pairs / singletons - I was waiting to hear from you if you had any thoughts about this? See my [comment above](https://github.com/ewels/MultiQC/pull/1174#pullrequestreview-404078317)

**Comment by noirot** [comment]
Rhhoooo ... sorry again !

**Comment by noirot** [comment]
It's so unituitive for me, I deleted the fork Rrrrrrrrhhh

**Comment by noirot** [comment]
I can redo it and integrate your comment I still have my code locally.

File: multiqc/modules/sickle/sickle.py (hunk 5 above)
Comment (noirot) [reply multiqc/modules/sickle/sickle.py]: No need to duplicate paired count.

**Comment by ewels** [comment]
Ah oops! Maybe I can still merge this using the command line, I’ll try that now..

**Comment by ewels** [comment]
Nope, it didn't like it on the command line sorry. This page still shows the code so we can copy + paste from that if necessary, but if you still have the code locally then that would be best :+1:

File: multiqc/modules/sickle/sickle.py (hunk 5 above)
Comment (ewels) [reply multiqc/modules/sickle/sickle.py]: Ok great - are they already 2 for each pair?

**Comment by ewels** [comment]
@noirot - never mind, I still had your changes on a branch locally, sorry I just noticed! So it's fine, I was able to merge locally and push, so everything is done.. 🎉

**Comment by ewels** [comment]
So this PR still shows _Closed_ instead of _Merged_, but despite this it is merged. You can see your commits on `master`: https://github.com/ewels/MultiQC/commits?author=noirot and Sickle is now listed on the [MultiQC homepage](https://multiqc.info/#supported-tools)...

So based on the fact that [you say we're ok with the pair read counts](https://github.com/ewels/MultiQC/pull/1174#discussion_r419279415), we're finished! Everything is done. Thank you! 🙌

**Comment by noirot** [comment]
Thank you! Now go for next tool ! :)

**Pull request closed without merging**